cli: Prepare login command for osemgrep invocation as default

zzeleznick opened the pull request:
## Description
This PR addresses some regression introduced with the login command and includes additional support for logging.

## Changes

### Adds
- Common config to `login` and `logout` subcommands to make `--experimental` flag not error out
- Debugging and trace information in create login token 

### Fixes
- Login handler not animating the loading spinner
- Login command appearing to hang due to aggressive sleep times
- `get_deployment_from_token` vs `get_scan_config_from_token`

## Test Plan
- [x] Verified login now works
- [x] Verified correct headers are sent using nc

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

zzeleznick committed:
fix what appears to be broken tests when run locally

zzeleznick committed:
simplify config fetching

zzeleznick committed:
fix login command: 1) use sleep appropriately to show spinner and not hang, 2) pass --experimental flag without the pain, 3) use deterministic anon id post-login

zzeleznick committed:
add client id with note

zzeleznick committed:
fix test payload

zzeleznick committed:
proactively update the privacy policy of the semgrep CLI but note the legal one is at https://semgrep.dev/privacy/

pabloest commented on PRIVACY.md:
```diff
@@ -193,21 +193,31 @@ Semgrep, Inc will:
-This ID is stored in the `~/.semgrep/settings.yml` file.
-If the ID disappears, the next run will generate a new one randomly.
-The Semgrep team uses this to answer the following questions:
+We will generate a new random UUID and save the ID locally to a `~/.semgrep/settings.yml` file
```
```suggestion
Semgrep generates a new random UUID and saves the ID locally to a `~/.semgrep/settings.yml` file
```

pabloest commented on PRIVACY.md:
```diff
@@ -193,21 +193,31 @@ Semgrep, Inc will:
 - > How many people use a given rule/ruleset/snippet?
 
-  This helps our security researchers assess their performance,
+  This enables our security researchers to assess their performance,
```
```suggestion
  This enables the Semgrep Security Research Team to assess their performance,
```

pabloest commented on PRIVACY.md:
```diff
@@ -193,21 +193,31 @@ Semgrep, Inc will:
-- > What percentage of users log in?"
+- > What percentage of users log in?
+
+  We use this to evaluate our success as we build new authenticated features for the Semgrep App.
```
```suggestion
  We use this to evaluate our success as we build new authenticated features for the Semgrep Cloud Platform.
```

pabloest commented on PRIVACY.md:
```diff
@@ -193,21 +193,31 @@ Semgrep, Inc will:
+1. **Randomization:** This ID is generated using a random algorithm (UUID V4) and does not contain any personally identifiable information (PII). Even with access to this ID, third-parties will not be able to associate this ID with a specific person.
+2. **Rotation:** This ID can be regularly reset, changed, or rotated, which limits entities to observe individual usage over an extended period of time.
+3. **Isolation:** This ID is only used for a specific purpose of analytics within a limited context of Semgrep and its product offerings.
+4. **Pseudonymization:** For logged in users who have accepted our terms of service and privacy policy, we will sanitize and transform (using a deterministic cryptographically secure hash) the PII used in generating a stable ID before sending data to our Semgrep, Inc backend and associated metrics collection endpoints.
```
```suggestion
4. **Pseudonymization:** For logged in users who have accepted Semgrep’s terms of service and privacy policy, we will sanitize and transform (using a deterministic cryptographically secure hash) the PII used in generating a stable ID before sending data to the Semgrep, Inc backend and associated metrics collection endpoints.
```

pabloest reviewed (commented):
copy edit suggestions

zzeleznick committed:
Apply suggestions from code review

updating the privacy policy with few copy edits, but will descope the stable uuid implementation from this PR and associated copy edits.

Co-authored-by: Pablo Estrada <pablo@r2c.dev>

zzeleznick committed:
revert stable uuidv5 and privacy policy changes

zzeleznick committed:
fix unused var

zzeleznick committed:
update comment

zzeleznick committed:
include cli id w/ header on both implementations

dlukeomalley commented on PRIVACY.md:
```diff
@@ -193,21 +193,23 @@ Semgrep, Inc will:
 - > How many people use a given rule/ruleset/snippet?
 
-  This helps our security researchers assess their performance,
+  This enables the Semgrep Security Research Team to assess their performance,
```
```suggestion
  This enables the Semgrep team to assess their performance,
```

dlukeomalley reviewed (approved):
Approving on the privacy clarifications. Defer to engineering on the implementation! 🎉

zzeleznick committed:
attempt merge

zzeleznick committed:
add doc suggestion

zzeleznick committed:
wrap-up merge

zzeleznick committed:
resolve payload diffs

zzeleznick committed:
resolve merged payload shapes
